discovery/aws/ec2.go

```diff
@@ -182,7 +183,7 @@ func NewEC2Discovery(conf *EC2SDConfig, logger log.Logger, metrics discovery.Dis
 	return d, nil
 }
 
-func (d *EC2Discovery) ec2Client(context.Context) (*ec2.EC2, error) {
+func (d *EC2Discovery) ec2Client(context.Context) (ec2iface.EC2API, error) {
```

Comment:
I think you can still return the concrete type here. Client code can decide to use a interface or the concrete type.

Reply:
I receive this when I change that back.
```
~/prometheus/discovery/aws $ go test -v
# github.com/prometheus/prometheus/discovery/aws [github.com/prometheus/prometheus/discovery/aws.test]
./ec2.go:188:10: cannot use d.ec2 (variable of type ec2iface.EC2API) as *ec2.EC2 value in return statement: need type assertion
./ec2.go:221:9: cannot use d.ec2 (variable of type ec2iface.EC2API) as *ec2.EC2 value in return statement: need type assertion
FAIL    github.com/prometheus/prometheus/discovery/aws [build failed]
```
I tried using type assertion and now it complains about functions are not implemented. Which is true as the ec2 has zillions of functions and I mocked/implemented only the ones which are called in the test.
```
ec2.EC2 does not implement ec2iface.EC2API (method AcceptAddressTransfer has pointer receiver)
```
Or maybe I misunderstood the error message. What was in your mind exactly? How to change the client code?

Reply:
Thank you. I marking this thread as resolved to keep the PR easy to read.